phylogeny/prototypeSelection.py

```diff
@@ -0,0 +1,106 @@
+"""
+Prototype selection
+
+Given a set of n elements (S) and their pairwise distances in terms of a
+dissimilarity matrix (DM) and a given number (k << n), find the sub-set (s)
+consisting of exactly k elements, such that s best represents the full set S.
+
+Here, we define "best represents" as maximizing the sum of distances between
+all points, i.e. sum {over a,b in S} DM[a,b]. This is our objective function.
+
+This problem is known to be NP-hard [1], thus we need to resort to heuristics.
+In the future, this module will implement several heuristics, whose quality can
+be measured by the objective function for each problem instance, since there is
+no global winner.
+For completeness, the exact but exponential algorithm is implemented, too.
+  "prototypeSelection_exhaustive"
+
+[1] Gamez, J. Esteban, François Modave, and Olga Kosheleva.
+    "Selecting the most representative sample is NP-hard:
+     Need for expert (fuzzy) knowledge."
+    Fuzzy Systems, 2008. FUZZ-IEEE 2008.
+"""
+
+from skbio.stats.distance import DistanceMatrix
+import itertools
+import numpy as np
+import scipy as sp
+
+
+def distanceSum(elements, dm: DistanceMatrix):
+    '''Compute the sum of pairwise distances for the given elements according to
+    the given dissimilarity matrix.
+
+    Parameters
+    ----------
+    elements: list or set
+        list or set of element IDs for which the sum of distances is computed
+    dm: skbio.stats.distance.DistanceMatrix
+        pairwise dissimilarity matrix.
+
+    Returns
+    -------
+    float:
+        the sum of all pairwise distances of dm for IDs in elements
+    '''
+
+    # some assertions
+    if type(dm) != DistanceMatrix:
+        raise TypeError('dm is not of type "DistanceMatrix"')
+    assert(len(elements) > 1)
+    assert(len(elements) <= len(dm.ids))
```

Comment:
Only support 3.5? What's QIIME 2 doing?

On Sep 2, 2016 10:23, "Stefan Janssen" <notifications@github.com> wrote:

> In phylogeny/prototypeSelection.py
> <https://github.com/biocore/WGS-HGT/pull/32#discussion_r77382119>:
>
> > +    elements: list or set
> > +        list or set of element IDs for which the sum of distances is computed
> > +    dm: skbio.stats.distance.DistanceMatrix
> > +        pairwise dissimilarity matrix.
> > +
> > +    Returns
> > +    -------
> > +    float:
> > +        the sum of all pairwise distances of dm for IDs in elements
> > +    '''
> > +
> > +    # some assertions
> > +    if type(dm) != DistanceMatrix:
> > +        raise TypeError('dm is not of type "DistanceMatrix"')
> > +    assert(len(elements) > 1)
> > +    assert(len(elements) <= len(dm.ids))
>
> arrrg. I just noticed that type annotation cause errors on the travis
> build for python 3.4
> What to do now? Ignore that problem or remove the beautiful type
> annotations?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/biocore/WGS-HGT/pull/32/files/f12a6fa86bd8fbfda95231d3b1a11d318a85014d#r77382119>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAc8suZTshdOdpsqFJY7fYvlHo-Ux8vuks5qmFuHgaJpZM4JyK-v>
> .
>